libevmasm/Inliner.cpp

```diff
@@ -0,0 +1,322 @@
+		 *
+		 * tag_1_out:
+		 * ...<potentially more jumps>...
+		 * JUMP // out, eventually
+		 *
+		 * tag_2:
+		 * ...<potentially more jumps>...
+		 * JUMP // out, eventually
+		 */
+		{{AssemblyItem::JumpType::IntoFunction, AssemblyItem::JumpType::IntoFunction}, AssemblyItem::JumpType::IntoFunction},
+		/*
+		 * PUSHTAG(main_out)
+		 * JUMP(f) // intoBlock: into
+		 * main_out:
+		 * ...
+		 *
+		 * f:
+		 * ...{A}...
+		 * JUMP // outOfBlock: out
+		 *
+		 * to
+		 *
+		 * PUSHTAG(main_out)
+		 * ...{A}...
+		 * JUMP // outOfBlock: ordinary
+		 * main_out:
+		 * ...
+		 *
+		 *
+		 */
+		{{AssemblyItem::JumpType::IntoFunction, AssemblyItem::JumpType::OutOfFunction}, AssemblyItem::JumpType::Ordinary},
+		/*
+		 * ...{A}...
+		 * JUMP(return) // intoBlock: out of
+		 * return:
+		 * ...{B}...
+		 * JUMP(tag_2) // outOfBlock: ordinary
+		 *
+		 * tag_2:
+		 * ...
+		 *
+		 * to
+		 *
+		 * ...{A}...
+		 * ...{B}...
+		 * JUMP(tag_2) // result: out of
+		 *
+		 * tag_2:
+		 * ...
+		 *
+		 *
+		 */
+		{{AssemblyItem::JumpType::OutOfFunction, AssemblyItem::JumpType::Ordinary}, AssemblyItem::JumpType::OutOfFunction},
+		/*
+		 * ...{A}...
+		 * JUMP(return) // intoBlock: out of
+		 * return:
+		 * ...{B}...
+		 * PUSH(tag_2_out)
+		 * JUMP(tag_2) // outOfBlock: into
+		 * tag_2_out:
+		 * ...{C}...
+		 *
+		 * tag_2:
+		 * ...<potentially more jumps>...
+		 * JUMP // out, eventually
+		 *
+		 * to
+		 *
+		 * ...{A}...
+		 * ...{B}...
+		 * PUSH(tag_2_out)
+		 * JUMP(tag_2) // result: ordinary
+		 * tag2_out:
+		 * ...{C}...
+		 *
+		 * tag_2:
+		 * ...<potentially more jumps>...
+		 * JUMP // out, eventually
+		 */
+		{{AssemblyItem::JumpType::OutOfFunction, AssemblyItem::JumpType::IntoFunction}, AssemblyItem::JumpType::Ordinary},
+		/*
+		 * ...{A}...
+		 * JUMP(return) // intoBlock: out of
+		 *
+		 * return:
+		 * ...{C}...
+		 * JUMP(return2) // outOfBlock: out of
+		 *
+		 * return2:
+		 * ...{D}...
+		 *
+		 * to
+		 *
+		 * ...{A}...
+		 * ...{C}...
+		 * JUMP(return2) // outOfBlock: out of (2x)
+		 *
+		 * return2:
+		 * ...{D}...
+		 *
+		 */
+		{{AssemblyItem::JumpType::OutOfFunction, AssemblyItem::JumpType::OutOfFunction}, AssemblyItem::JumpType::OutOfFunction},
+	};
+	return jumpTypeMapping.at(make_pair(_intoBlock, _outOfBlock));
+}
+}
+
+void Inliner::optimise()
+{
+	std::map<u256, ranges::span<AssemblyItem const>> inlinableBlocks = determineInlinableBlocks(m_items);
+
+	if (inlinableBlocks.empty())
+		return;
+
+	AssemblyItems newItems;
+	for (auto it = m_items.begin(); it != m_items.end(); ++it)
+	{
+		AssemblyItem const& item = *it;
+		if (next(it) != m_items.end())
+		{
+			AssemblyItem const& nextItem = *next(it);
+			if (item.type() == PushTag && nextItem == Instruction::JUMP)
+				if (auto const *inlinableBlock = util::valueOrNullptr(inlinableBlocks, item.data()))
+				{
+					newItems += *inlinableBlock;
+					newItems.back().setJumpType(determineJumpType(nextItem.getJumpType(), inlinableBlock->back().getJumpType()));
```

Comment:
I guess the main problem is when the breaking instruction is not a jump instruction - you cannot really set a jump type on a call opcode.

Reply:
It will always be a jump (as I just explained on gitter and as seen in determineInlinableBlocks) :-).